### PR TITLE
Fusion plant features for THERM

### PR DESCRIPTION
**Pull request by cfe316**
Description of the fusion module
--------------------

FUSION is, more or less, a few extra operational constraints plus an expression for parasitic (recirculating) power.

* Fusion plants are set of extra variables, expressions, and constraints that "bolt on" to a THERM=1 plant.
  * This is similar to how the Maintenance module is a "bolt-on" to a THERM=1 plant.
  * Fusion and Maintenance are fully cross-compatible.
  
* This PR does not include multi-hour thermal storage (as my _Joule_ paper had), but it does keep that future application in mind. I think the best picture is that it _does_ include sub-hourly thermal storage (perhaps 10 or 20 minutes worth) for the time between pulses.

* A fusion plant has three types of parasitic power, expressed in _five_ parameters
    * Passive parasitic power is on at a constant rate all the time (but can be specified as being lower during maintenance.) This models vacuum pumps, cryoplant, tritium handling systems etc which are always on.
    * Active parasitic power is on when a 'pulse' is on. This models RF or NBI heating.
    * Start energy is extra energy required during the hour a pulse starts. This models fast solenoid swings or extra heating power.
    * Start _power_ models a very large, but brief peak power level (e.g. 300 MW for a minute) during startup. This affects the CapacityReserveMargin policy, but not overall power balance.
    * `parasitic_passive_maintenance_remaining` reduces passive parasitic power during maintenance.
    
* Two new variables vPulseStart and vPulseUnderway keep track of pulses. These are roughly analogous to vSTART and vCOMMIT. 
    * One can set a "dwell time" that occurs at the start of a pulse, to model the 10-15 minutes time between pulses. During this time the fusion core can't generate heat.
    * It's possible to specify a maximum pulse length, e.g. 2 hours for something like EU-DEMO. This makes the reactor need to vPulseStart every other hour.
    * It's also possible to model a maximum number of pulses per annum, if you want to do that.
    
* The fusion variables interact with the standard THERM variables through `vP` and `vCOMMIT`:
    * The core can't pulse unless the plant is `vCOMMIT`'d.
    * `vP <= vPulseUnderway - dwell_time * vPulseStart`
    
* Standard constraints on vP apply
    * Min_power
    * Max_power (pP_Max)
    * Ramp up / Ramp_down
    * Minimum Up_time / down_time

* There are no additional constraints for something like the ramp rate of the core: fusion physics is much faster than 1h.
       
* The net power of a fusion plant is `vP[t] - eFusionParasiticTotal[t]`.

The interface
--------------------
* No new input files.
* 9 + 1 new, optional columns in Thermal.csv
  1. FUSION (could be eliminated and just used if non-default values in the other columns were detected?)
  2. Parasitic_Passive
  3. Parasitic_Active
  4. Parasitic_Start_Energy
  5. Dwell_Time
  6. Max_Up_Time
  7. Max_Starts
  8. Parasitic_Start_Power
  9. Parasitic_Passive_Maintenance_Remaining (only required if the plant also has Maintenance turned on).
  10. Max_FPY_per_year

One new output-writing boolean setting `WriteFusion`, controlling
two new output files:
* fusion_net_capacity_factor.csv
* fusion_pulse_starts.csv

New column in power_balance.csv "Fusion parasitic power"

Summary / implementation progress
---------------

- [x] Functions to apply Fusion to resources (or components of resources, e.g. the thermal core for future thermal+storage devices).
- [x] Interoperable with Maintenance formulation
- [x] Adjustments to zonal power balance for parasitic (recirculating) power

- [x] Add adjustments to capacity reserve margin
- [x] Add adjustments to Reserves
- [x] Add adjustments to ESR policy.
- Hook up to specialized output files
  - [x] capacityfactor.csv: via new output file `fusion_net_capacity_factor.csv`
  - [x] CapacityValue.csv: via modifications to `thermal_plant_effective_capacity`.
  - [x] charge.csv: parasitic power is charging.
  - [x] ChargingCost.csv: ditto
  - [x] ESR_Revenue
  - [x] power_balance.csv : parasitic power in new fusion column (since it's certainly not "storage charge")
  - [x] ReserveMarginRevenue : incorporates parasitic power.
  
 - [x] Documentation 
   - [x] Basic description
   - [x] How to use
   - [x] Full mathematical formulation 

- [x] Test case (in lieu of example)

## What type of PR is this? (check all applicable)

- [x] Feature
- [ ] Bug Fix
- [ ] Documentation Update
- [ ] Code Refactor
- [ ] Performance Improvements

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and .md files under /docs/src have been updated if necessary.
- [x] The latest changes on the target branch have been incorporated, so that any conflicts are taken care of before merging. This can be accomplished either by merging in the target branch (e.g. 'git merge develop') or by rebasing on top of the target branch (e.g. 'git rebase develop'). Please do not hesitate to reach out to the GenX development team if you need help with this.
- [x] Code has been tested to ensure all functionality works as intended.
- [x] CHANGELOG.md has been updated (if this is a 'notable' change).
- [x] I consent to the release of this PR's code under the GNU General Public license.

## How this can be tested

- if there's no FUSION column or FUSION=0 there should be no change.
- if FUSION=1, but all fusion-columns have their default values, there should be no change.

## Post-approval checklist for GenX core developers
After the PR is approved

- [ ] Check that the latest changes on the target branch are incorporated, either via merge or rebase
- [ ] Remember to squash and merge if incorporating into develop

### CODE REVIEW

File: src/model/resources/thermal/thermal_commit.jl [hunk 1]
```diff
@@ -351,7 +368,7 @@ function maintenance_formulation_thermal_commit!(EP::Model, inputs::Dict, setup:
     maint_freq(y) = Int(floor(by_rid(y, :maintenance_cycle_length_years)))
     maint_begin_cadence(y) = Int(floor(by_rid(y, :maintenance_begin_cadence)))
 
-    integer_operational_unit_committment = setup["UCommit"] == 1
+    integer_operational_unit_commitment = setup["UCommit"] == 1
```
Comment (cfe316) [inline src/model/resources/thermal/thermal_commit.jl:371]: This is an unrelated spelling correction.

File: src/write_outputs/capacity_reserve_margin/effective_capacity.jl [hunk 2]
```diff
@@ -1,6 +1,6 @@
 @doc raw"""
-	thermal_plant_effective_capacity(EP::Model,
-	                                 inputs::Dict,
+    thermal_plant_effective_capacity(EP::Model,
+                                     inputs::Dict,
                                      resources::Vector{Int},
                                      capres_zone::Int,
                                      timesteps::Vector{Int})::Matrix{Float64})
```
Comment (cfe316) [inline src/write_outputs/capacity_reserve_margin/effective_capacity.jl:6]: This is from formatting with JuliaFormatter. I'm not sure why this didn't get picked up before.

File: src/write_outputs/write_power_balance.jl [hunk 3]
```diff
@@ -68,7 +72,7 @@ function write_power_balance(path::AbstractString, inputs::Dict, setup::Dict, EP
         if !isempty(intersect(resources_in_zone_by_rid(gen, z), VRE_STOR))
             VS_ALL_ZONE = intersect(resources_in_zone_by_rid(gen, z), inputs["VS_STOR"])
 
-            # if ELECTROLYZER is empty, increase indices by 1
+            # if ELECTROLYZER is not empty, increase indices by 1
```
Comment (cfe316) [inline src/write_outputs/write_power_balance.jl:75]: This is an unrelated textual correction.

File: src/model/resources/fusion/fusion.jl [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+)
+    from_model(f::Function) = EP[Symbol(f(resource_component))]
+
+    ePassive = from_model(fusion_parasitic_passive_name)
+    eActive = from_model(fusion_parasitic_active_name)
+    eStartEnergy = from_model(fusion_parasitic_start_energy_name)
+
+    total_parasitic = Symbol(fusion_parasitic_total_name(resource_component))
+
+    EP[total_parasitic] = ePassive + eActive + eStartEnergy
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion.jl:503]: If possible, create expressions inside JuMP macros ([Ref](https://jump.dev/JuMP.jl/stable/tutorials/getting_started/performance_tips/#Use-macros-to-build-expressions))

File: src/model/resources/fusion/fusion.jl [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+    # Startup energy, taken from the grid every time the core starts up
+    EP[start_energy] = @expression(EP,
+        [t in 1:T],
+        pulsestart[t]*η*parasitic_start_energy_fraction)
+
+    # Startup power, required margin on the grid when the core starts
+    EP[pulse_start_power] = @expression(EP,
+        [t in 1:T],
+        pulsestart[t]*η*pulse_start_power_fraction)
+    union!(inputs[FUSION_PULSE_START_POWER], (pulse_start_power,))
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion.jl:486]: Maybe move this `Set` update up, similar to what you did in `fusion_pulse_variables!`, to improve readability

File: src/model/resources/fusion/fusion_maintenance.jl [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    end
+end
+
+# lower level function
+function _fusion_maintenance_parasitic_power_adjustment!(
+        EP::Model, resource_component, reactor::FusionReactorData)
+    passive = reactor.parasitic_passive_fraction
+    reduction_factor = reactor.maintenance_remaining_parasitic_power_fraction
+    η = reactor.eff_down
+    cap_size = reactor.component_size
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion_maintenance.jl:21]:
```suggestion
    component_size = reactor.component_size
```
to follow same convention as in `fusion.jl`

And the same for `get_from_model` -> `from_model`

File: src/model/resources/fusion/fusion_maintenance.jl [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    reduction = passive * (1 - reduction_factor)
+
+    get_from_model(f::Function) = EP[Symbol(f(resource_component))]
+
+    ePassiveParasitic = get_from_model(fusion_parasitic_passive_name)
+    vMDOWN = get_from_model(maintenance_down_name)
+
+    eReduction = -reduction * η * cap_size * vMDOWN
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion_maintenance.jl:30]: Same as https://github.com/GenXProject/GenX.jl/pull/743#discussion_r1750797163

File: src/model/resources/thermal/thermal.jl [hunk 8]
```diff
@@ -34,9 +36,18 @@ function thermal!(EP::Model, inputs::Dict, setup::Dict)
             for y in THERM_ALL))
         add_similar_to_expression!(EP[:eCapResMarBalance], eCapResMarBalanceThermal)
 
-        MAINT = ids_with_maintenance(gen)
         if !isempty(intersect(MAINT, THERM_COMMIT))
             thermal_maintenance_capacity_reserve_margin_adjustment!(EP, inputs)
         end
+
+        if !isempty(intersect(FUSION, THERM_COMMIT))
```
Comment (lbonaldo) [inline src/model/resources/thermal/thermal.jl:43]: Is it possible for FUSION plant indices not to be in THERM_COMMIT?

File: src/model/resources/fusion/fusion_policy_adjustments.jl [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+#################################
+
+function fusion_parasitic_power_adjust_energy_share_requirement!(EP, inputs)
+    eESR = EP[:eESR]
+    nESR = inputs["nESR"]
+    gen = inputs["RESOURCES"]
+    FUSION = ids_with(gen, fusion)
+
+    for y in FUSION, p in 1:nESR
+        esr_derating = get(gen[y], Symbol("ESR_" * string(p)), 0)
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion_policy_adjustments.jl:133]: Why not using `esr()`?

File: src/model/resources/fusion/fusion.jl [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+    # Maximum thermal power generated by core y at hour y <= Max power of committed
+    # core minus power lost from down time at startup
+    @constraint(EP,
+        [t in 1:T],
+        power_like[r_id,
+            t]<=ePulseUnderway[t] -
+                _fusion_dwell_avoided_operation(dwell_time, ePulseStart[t]))
+end
+
+function fusion_max_fpy_per_year_constraint!(
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion.jl:407]: You could move this function after the following two functions to align with the workflow in `fusion_formulation_thermal_commit!`

File: src/model/resources/thermal/thermal.jl (hunk 8 above)
Comment (cfe316) [reply src/model/resources/thermal/thermal.jl:43]: Not right now, but we could consider a THERM_NOCOMMIT fusion later.

File: docs/src/Model_Reference/Resources/fusion.md [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+The value must be between $0$ and $1$.
+The default is $0$; in this case there is no parasitic power during scheduled maintenance periods.
+With a value of $1$ the parasitic power does not decrease during maintenance.
+With a value of $0.5$ it is reduced by half, and so on.
+
+## Outputs produced
+
+Use of the fusion module leads to two new outputs files, and changes to existing output files.
+
+- `fusion_pulse_starts`. This file is similar to `start/commit/shut.csv` for unit committment.
```
[the remark this answers is not in the document]
Comment (cfe316) [reply docs/src/Model_Reference/Resources/fusion.md:129]: fixed.

File: src/model/resources/fusion/fusion.jl [hunk 12]
```diff
@@ -0,0 +1,545 @@
...
+        parasitic_start_energy_fraction,
+        pulse_start_power_fraction,
+        maintenance_remaining_parasitic_power_fraction,
+        eff_down,
+        dwell_time,
+        max_pulse_length,
+        max_starts,
+        max_fpy_per_year
+    )
+end
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl:194]: Would that require Julia 1.9? Or is it okay to use`Base.@kwdef`, but not `@kwdef` ?

File: src/model/resources/fusion/fusion_policy_adjustments.jl (hunk 9 above)
Comment (cfe316) [reply src/model/resources/fusion/fusion_policy_adjustments.jl:133]: thanks

File: src/model/resources/fusion/fusion.jl [hunk 13]
```diff
@@ -0,0 +1,545 @@
...
+    for var in (FUSION_PARASITIC_POWER, FUSION_PULSE_START_POWER)
+        if var ∉ keys(dict)
+            dict[var] = Set{Symbol}()
+        end
+    end
+    var = FUSION_COMPONENT_ZONE
+    if var ∉ keys(dict)
+        dict[var] = Dict{Int, Set{AbstractString}}()
+    end
+end
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl:67]: This doesn't work because when you call 

`ensure_records!(dict, [FUSION_PULSE_START, FUSION_PULSE_UNDERWAY], Set{Symbol}())`

`Set{Symbol}()` gets evaluated only once, not once per key.

Maybe I could solve this by calling
`ensure_records!( ..., ..., Set{Symbol})` where the function has
`dict[key] = default_value()`
?

What do you think about that?

File: src/model/resources/fusion/fusion.jl [hunk 14]
```diff
@@ -0,0 +1,545 @@
...
+        parasitic_active_fraction = parasitic_active,
+        parasitic_start_energy_fraction = start_energy,
+        pulse_start_power_fraction = start_power,
+        eff_down = eff_down,
+        dwell_time = dwell_time,
+        max_pulse_length = max_pulse_length,
+        max_starts = max_starts,
+        maintenance_remaining_parasitic_power_fraction = parasitic_maint,
+        max_fpy_per_year = max_fpy_per_year)
+end
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl:209]: I'm not sure what you mean here. How would you write `function FusionReactorData(r::Thermal)` ?

File: docs/src/Model_Reference/Resources/fusion.md [hunk 15]
```diff
@@ -0,0 +1,289 @@
+# Fusion
+_Added in v0.4.2_
+
+Fusion power plants could be a future source of clean, firm electricity.
+Most proposals for fusion plants use a thermal cycle, where the energy of fusion heats a working fluid.
+The working fluid passes through a turbine connected to a generator.
+These plants would be broadly similar to fission plants.
+They might also have some fusion-specific operational behaviors, which are implemented in this module.
+
+This fusion module is implemented **only** for thermal plants with unit commitment (THERM=1).
```
[the remark this answers is not in the document]
Comment (cfe316) [reply docs/src/Model_Reference/Resources/fusion.md:10]: I think it will work but it will not give useful results. It will completely ignore the fusion-specific parts of the model.

I think there is a reasonable relaxation of fusion to a `THERM_NO_COMMIT` model: basically a fusion plant with its pulse behavior averaged out over time. This would be equivalent to a standard thermal plant + an extra load proportional to fusion capacity.

I could do this but I'm feeling lazy about it. And it would need some additional conditions on the back-end / outputs. Maybe let's chat about it?

[the remark this answers is not in the document]
Comment (cfe316) [reply docs/src/Model_Reference/Resources/fusion.md:10]: Maybe at the very least the user should get an error (even more than a warning) when running fusion with ucommit=2 or model = 2. An error is called for because the output would be different from what the modeler would reasonably expect.

File: src/model/resources/fusion/fusion.jl (hunk 4 above)
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl]: done.

File: src/model/resources/fusion/fusion.jl (hunk 5 above)
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl]: done

File: src/model/resources/fusion/fusion.jl [hunk 16]
```diff
@@ -0,0 +1,545 @@
...
+        capacity::Symbol
+)
+    T = inputs["T"]
+    ω = inputs["omega"]
+    p = inputs["hours_per_subperiod"]
+
+    start_name = fusion_pulse_start_name(resource_component)
+    underway_name = fusion_pulse_underway_name(resource_component)
+    start = Symbol(start_name)
+    underway = Symbol(underway_name)
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl]: done

File: docs/src/Model_Reference/Resources/fusion.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (lbonaldo) [reply docs/src/Model_Reference/Resources/fusion.md:10]: Yes, I agree with adding a check to ensure that a model with fusion resources has ucommit > 0 in the settings. If the user wants to run it without ucommit, they can simply remove the fusion plants. In principle, we could handle this internally, but it would be less explicit.

File: docs/src/Model_Reference/Resources/fusion.md [hunk 17]
```diff
@@ -0,0 +1,289 @@
...
+
+- `charge`: fusion parasitic power is reported in this file.
+
+## Mathematical formulation of fusion module
+
+### Pulse start and pulse status
+There are two new variables, representing the number of pulse starts ($\chi^{pulse}_{y,z,t}$) and pulses underway ($\nu^{pulse}_{y,z,t}$).
+These are similar to the conventional unit commitment start $\chi$ and commit status $\nu$.
+
+`TODO: tidy up math presentation?`
```
Comment (lbonaldo) [inline docs/src/Model_Reference/Resources/fusion.md:154]: Is this still a work in progress? The math part looks great.

File: src/model/resources/fusion/fusion.jl (hunk 13 above)
[the remark this answers is not in the document]
Comment (lbonaldo) [reply src/model/resources/fusion/fusion.jl:67]: Sorry, there’s an error, `keys_to_check` should be a `Vector{String}`. I’m not sure I fully understood your comment, isn’t this the expected behavior?
```
julia> function ensure_records!(dict::Dict, keys_to_check::Vector{String}, default_value)
           for key in keys_to_check
               if key ∉ keys(dict)
                   dict[key] = default_value
               end
           end
       end
ensure_records! (generic function with 2 methods)

julia> d = Dict()
Dict{Any, Any}()

julia> ensure_fusion_pulse_variable_records!(d)

julia> d
Dict{Any, Any} with 2 entries:
  "FusionPulseUnderwayVariables" => Set{Symbol}()
  "FusionPulseStartVariables"    => Set{Symbol}()

julia> ensure_fusion_expression_records!(d)

julia> d
Dict{Any, Any} with 5 entries:
  "FusionComponentZones"             => Dict{Int64, Set{AbstractString}}()
  "FusionParasiticPowerExpressions"  => Set{Symbol}()
  "FusionPulseUnderwayVariables"     => Set{Symbol}()
  "FusionPulseStartPowerExpressions" => Set{Symbol}()
  "FusionPulseStartVariables"        => Set{Symbol}()
```

[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl:67]: The problem is that there's only one `Set` object shared by multiple keys. So if I add an element to the `Set` under `d["FusionParasiticPowerExpressions"]` it also appears under `d["FusionPulseStartPowerExpressions"]`.

File: src/model/resources/fusion/fusion.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (lbonaldo) [reply src/model/resources/fusion/fusion.jl:209]: I was thinking something like this:
```
function FusionReactorData(r::Thermal)::FusionReactorData

    return FusionReactorData(
        component_size = Float64(cap_size(r)),
        parasitic_passive_fraction = Float64(get(r, :parasitic_passive, default_zero)),
        parasitic_active_fraction = Float64(get(r, :parasitic_active, default_zero)),
        parasitic_start_energy_fraction = Float64(get(r, :parasitic_start_energy, default_zero)),
        pulse_start_power_fraction = Float64(get(r, :parasitic_start_power, default_zero)),
        eff_down = 1.0,
        dwell_time = Float64(get(r, :dwell_time, default_zero)),
        max_pulse_length = get(r, :max_up_time, -1),
        max_starts = get(r, :max_starts, -1),
        maintenance_remaining_parasitic_power_fraction =  Float64(get(r, :parasitic_passive_maintenance_remaining, default_zero)),
        max_fpy_per_year = Float64(get(r, :max_fpy_per_year, -1))
end
```
but it doesn't change much probably

File: src/load_inputs/load_resources_data.jl [hunk 18]
```diff
@@ -414,6 +414,46 @@ function check_maintenance_applicability(r::AbstractResource)
...
+        push!(error_strings, e)
+    end
+
+    if setup["UCommit"] == 0
+        e = string("Resource ", resource_name(r), " has :fusion = ", value, ".\n",
+                   "Use of the fusion module requires the setting UCommit > 0.\n",
+                   "Contact the fusion module maintainers (Jacob Schwartz) if you are interested\n",
+                   "in running a fusion case without unit commitment.")
+        push!(error_strings, e)
+    end
```
Comment (lbonaldo) [inline src/load_inputs/load_resources_data.jl:452]: Awesome, thank you @cfe316!

File: src/model/resources/fusion/fusion.jl [hunk 19]
```diff
@@ -0,0 +1,493 @@
...
+    FUSION_PARASITIC_POWER in keys(dict)
+end
+
+#######################################
+# Define a data structure for a reactor
+#######################################
+
+# Base.@kwdef could be used if we enforce Julia >= 1.9
+# That would replace need for the keyword-argument constructor below
+Base.@kwdef struct FusionReactorData
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion.jl:153]: `Base.@kwdef` should be available in Julia versions 1.1 and later. However, it needs to be fully qualified because it wasn't exported until Julia 1.9.

File: src/model/resources/fusion/fusion_policy_adjustments.jl [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+        maint_adj = thermal_maintenance_and_fusion_capacity_reserve_margin_adjustment(
+            EP, inputs, y, capres_zone, timesteps)
+        add_to_expression!.(total_adj, maint_adj)
+    end
+
+    return total_adj
+end
+
+# alias for better parallelism in effective_capacity.jl
+thermal_fusion_capacity_reserve_margin_adjustment = fusion_capacity_reserve_margin_adjustment
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion_policy_adjustments.jl:91]:
```suggestion
const thermal_fusion_capacity_reserve_margin_adjustment = fusion_capacity_reserve_margin_adjustment
```
to improve performance

File: src/write_outputs/write_maintenance.jl [hunk 21]
```diff
@@ -7,22 +7,42 @@ function write_simple_csv(filename::AbstractString, header::Vector, matrix)
...
-function write_maintenance(path::AbstractString, inputs::Dict, EP::Model)
+function prepare_maintenance_downvars_matrix(EP::Model, inputs::Dict)
     downvars = maintenance_down_variables(inputs)
-    write_timeseries_variables(EP, downvars, joinpath(path, "maint_down.csv"))
+    return prepare_timeseries_variables(EP, inputs, downvars, "vMDOWN_")
+end
+
+function write_maintenance(path::AbstractString, inputs::Dict, setup::Dict, EP::Model)
+    set, data = prepare_maintenance_downvars_matrix(EP, inputs)
+    gen = inputs["RESOURCES"][set]
```
Comment (lbonaldo) [inline src/write_outputs/write_maintenance.jl:45]:
```suggestion
```

File: src/write_outputs/write_fusion.jl [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+
+###############
+# writing data
+###############
+
+# NB: there is no 'write_fusion_parasitic_power'; it is in 'charge.csv'.
+
+function write_fusion_pulse_starts(
+        path::AbstractString, inputs::Dict, setup::Dict, EP::Model)
+    myret = prepare_fusion_pulse_starts(EP, inputs)
```
Comment (lbonaldo) [inline src/write_outputs/write_fusion.jl:38]:
```suggestion
```

File: src/write_outputs/write_fusion.jl [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+    set, data = prepare_fusion_pulse_starts(EP, inputs)
+    gen = inputs["RESOURCES"][set]
```
Comment (lbonaldo) [inline src/write_outputs/write_fusion.jl:40]:
```suggestion
```

File: src/model/resources/fusion/fusion_policy_adjustments.jl [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+#################################
+
+function fusion_parasitic_power_adjust_energy_share_requirement!(EP, inputs)
+    eESR = EP[:eESR]
+    nESR = inputs["nESR"]
+    gen = inputs["RESOURCES"]
+    FUSION = ids_with(gen, fusion)
+
+    for y in FUSION, p in 1:nESR
+        esr_derating = esr(gen[y], p)
```
Comment (lbonaldo) [inline src/model/resources/fusion/fusion_policy_adjustments.jl:133]:
```suggestion
        esr_derating = esr(gen[y], p)
```
```suggestion
        esr_derating = esr(gen[y], p)
```
```suggestion
        esr_derating = esr(gen[y], tag = p)
```

**Comment by lbonaldo** [comment]
I’ve completed the review and added a few more comments. Apart from those, it looks ready to me. Outstanding work, thanks again @cfe316!

File: src/model/resources/fusion/fusion.jl (hunk 12 above)
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl]: done.

File: src/model/resources/fusion/fusion.jl [hunk 25]
```diff
@@ -0,0 +1,545 @@
...
+        maintenance_remaining_parasitic_power_fraction = parasitic_maint,
+        max_fpy_per_year = max_fpy_per_year)
+end
+
+#######################################
+# Compute reactor properties
+#######################################
+
+#TODO consider moving these to resources.jl? Or get rid of them in favor of direct resource
+# access?
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl]: done.

File: docs/src/Model_Reference/Resources/fusion.md (hunk 17 above)
Comment (cfe316) [reply docs/src/Model_Reference/Resources/fusion.md]: done.

File: src/model/resources/fusion/fusion.jl (hunk 19 above)
Comment (cfe316) [reply src/model/resources/fusion/fusion.jl:153]: done.

File: src/model/resources/fusion/fusion_policy_adjustments.jl (hunk 20 above)
Comment (cfe316) [reply src/model/resources/fusion/fusion_policy_adjustments.jl]: done.

File: src/write_outputs/write_maintenance.jl [hunk 26]
```diff
@@ -7,22 +7,42 @@ function write_simple_csv(filename::AbstractString, header::Vector, matrix)
...
+    mat = Matrix(reduce(hcat, extract.(v))') # matrix of size length(syms) x T
+    return r_id, mat
 end
 
-function write_timeseries_variables(EP, set::Set{Symbol}, filename::AbstractString)
-    df = prepare_timeseries_variables(EP, set)
-    write_simple_csv(filename, df)
+function strip_common_prefix(v::Vector{Symbol}, prefix::AbstractString)
+    s = string.(v)
+    len = length(prefix)
```
[the remark this answers is not in the document]
Comment (cfe316) [reply src/write_outputs/write_maintenance.jl]: done. Thanks for catching.

File: src/write_outputs/write_maintenance.jl (hunk 21 above)
Comment (cfe316) [reply src/write_outputs/write_maintenance.jl]: done.

**Review by lbonaldo — APPROVED** [review]
